Make shell component serialisation plugable

jsgilmore opened the pull request:
This pull request makes multilang components plugable (issue #373) and relates to issue #654, in that it is now possible to implement protocol buffer serialisation.

Changes in this pull request:
Moved multilang classes to multilang directory.
Added BoltMsg, ShellMSg and SpoutMsg objects. These objects are used when communicating with a serializer, instead of raw JSON.
All JSON serialisation has been moved to a new JsonSerializer.
A serialiser is selected when creating a new ShellBolt or ShellSpout, by passing it in to the constructor. A default option of no serialiser automatically selects the built-in JsonSerializer.

Commit:
0.9.0-wip3

Commit:
multischeme.

Commit:
version bump.

Commit:
fix typo.

Commit:
multischeme related changes

miguno committed:
Completely refactor Rolling Count example

The new implementation separates the various concerns formerly mixed up in the
example code.  Also, we are now using tick tuples (introduced in Storm 0.8)
instead of spawning manual threads for carrying out periodical tasks.

Lastly, we add 192 unit tests for the new implementation that brings the test
coverage for the Rolling Count from 0% to almost 100%.

Note: Adding those unit tests required changes to the build (m2-pom.xml),
notably new test dependencies and moving the existing Java code from src/jvm/*
to src/jvm/main/*.  The latter was required so that the test runner triggered
by Maven can tell code (src/jvm/main) and tests (src/jvm/test) apart.

miguno committed:
Move code back to src/jvm and place unit tests into test/jvm

nathanmarz committed:
Merge pull request #27 from miguno/master

Complete refactoring of the Rolling Count example

Commit:
fix project.clj

Commit:
fix project.clj

miguno committed:
SlidingWindowCounterTest: add missing @Test annotation to unit test

nathanmarz committed:
Merge pull request #28 from miguno/master

SlidingWindowCounterTest: add missing @Test annotation to unit test

Commit:
version bump 0.9.0-wip5

Commit:
removed twitter repositories from project clj

Commit:
bumped version, upgraded storm dep

Commit:
lein2 upgrade
added kafkaspout lag to transactionaltridentkafkaspout

Commit:
include offset when logging about committing offsets

Commit:
Merge branch 'master' of github.com:nathanmarz/storm-contrib

trevor committed:
update to storm 0.8.2

Commit:
un-complect wiping zeroes and wiping a slot

Yes, we iterate through the list twice, but we no longer have
a function that does two things.

Commit:
wipe zeros *before* wiping slots, ensure that zeros get reported.

This commit is best explained by describing the behavior of
RollingTopWords in its absence.

Let's say the word "Bieber" appears in the tweet stream at a very high
rate for half an hour, and then is *never heard again*.

This will insert the word "Bieber" into our rankings object downstream.

Note that there is nothing to *remove* the word "Bieber" from the
rankings object downstream. Let us assume that, during its half-hour,
the word "Bieber" appears *more than five times as often* as other
highly-ranked words. Then the very *last* "Bieber" report (in which
"Bieber" has only appeared for the first minute of the five-minute
window) will still be able to hold a place in the rankings. Objects are
only lowered in the rankings by being outranked, or by *appearing again*
with a lower score.

*With* this change, we wipe zeros, and *then* wipe a slot. This means
that when we wipe the slot containing the last Bieber mentions, we
don't at that moment remove "Bieber" from our map. The *next* report
will include "Bieber" with a score of zero, removing "Bieber" from the
ranking object downstream.

Commit:
fixed bug. lastmeta data may be null in certain cases

nathanmarz committed:
Merge pull request #33 from MichaelBlume/emit-zeros

Ensure zeros get reported.

nathanmarz committed:
Merge pull request #32 from trevor/master

update to storm 0.8.2

Commit:
fixed concurrent modification exception

Commit:
bump 0.9.0-wip15e-scala292

Commit:
fix unit tests to match new behavior

Commit:
fixed possible issue where getValueAndReset is called before refreshPartitions is ever called

Commit:
bump 0.9.0-wip15f-scala292

Commit:
track average and max kafka request latencies

Commit:
0.9.0-wip16-scala292

Commit:
forgot MaxMetric

Commit:
fixed NPE in MaxMetric. bumped version 0.9.0-wip16a-scala292

nathanmarz committed:
upgrade to lein 2

nathanmarz committed:
Merge pull request #35 from MichaelBlume/fix-tests

fix unit tests to match new behavior

Commit:
add api latency metric to KafkaSpout

Commit:
bump 0.9.0-wip16b-scala292

wurstmeister committed:
Initial commit

Commit:
clone of nathanmarz/storm-contrib/storm-kafka

wurstmeister committed:
updated kafka installation instructions

wurstmeister committed:
fixed kafka 0.8 compilation issues

wurstmeister committed:
updated zookeeper dynamic broker detection

* split broker hosts into separate classes
* added test topology

wurstmeister committed:
extracted Coordinator to avoid code duplication

wurstmeister committed:
updated dynamic leader detection

* refactored partition handling

wurstmeister committed:
added test for reading partition leader

* moved to maven

wurstmeister committed:
moved to maven

wurstmeister committed:
Refactored emitters

* extracted emitter code to single class
* downgraded curator version to avoid conflicts

wurstmeister committed:
added support for client id

* removed unused leiningen project file

wurstmeister committed:
Update to kafka 0.8 beta1

* updated pom and build instructions

wurstmeister committed:
fixed formatting

wurstmeister committed:
fixed formatting

wurstmeister committed:
updated to storm 0.9.0-wip21

wurstmeister committed:
added license

wurstmeister committed:
fixed infinite loop in KafkaSpout

wurstmeister committed:
renamed GlobalPartitionId

wurstmeister committed:
removed JDK 7 dependency

miguno committed:
ISSUE-36: equals() should compare against input parameter

miguno committed:
storm-starter actually uses leiningen 2.0 now

Commit:
RankableObjectWithFields implements Serializable to fix NotSerializableException errors

Commit:
Explain how to run unit tests with Maven

Commit:
README: Use 120 chars max per line

Commit:
Reformat README

Commit:
Update README

Commit:
Restructure README and add more docs

Commit:
Update main heading

Commit:
Add table of contents

Commit:
Update Getting Started section

Commit:
PR-3: Add step-by-step instructions for the curious but unsavvy

Most of the original pull request (dated two years ago) was added to the
README.

See https://github.com/nathanmarz/storm-starter/pull/3 for the original
pull request.

Commit:
AbstractRankerBolt: use log level DEBUG instead INFO so that the bolt is more silent on the logging side

Commit:
RollingCountBolt: use log level DEBUG to tell that a tick tuple was received

Commit:
Rankings: make updateWith() thread-safe

This fixes a ConcurrentModificationException that could be triggered when
using log4j to log Rankings#toString() from within a bolt's execute()
method.  Apparently a separate thread is used by log4j to perform the
logging, whose execution path could interweave with the bolt's own
thread.

Even though this is not necessarily a deficiency of the Rankings class
(which does not claim to be thread-safe), we still decided to make
Rankings#updateWith() thread-safe so that users do not run into errors
when using log4j & Co. for logging purposes.  In other words, we decided
to add this fix for the convenience of our users.

Commit:
project.clj: add test dependencies of the Java code in storm-starter

Commit:
Reformat Java code to use 2 instead of 4 spaces (to match Clojure style)

Commit:
Rankings: add pruneZeroCounts() method

Commit:
Rankings: use private helper method to implement getRankings() for cleaner code

Commit:
project.clj: add explicit :test-paths setting so that Java IDE's can
properly match Java package names and directory names for test classes

Commit:
TotalRankingsBolt: prune zero count rankings after each emit

Commit:
RankableObjectWithFields: update Java docs

Commit:
SlidingWindowCounter: update Java docs

Commit:
Shell component serialisation is now plugable with a default JSON serialiser being used. Moved multilang classes to multilang directory.

xumingming commented:
change the package of `ShellBolt` makes this PR a breaking change: user code need to be modified to use the new package, is it really necessary?

jsgilmore commented:
It's not necessary, no. It was an effort to combine code that (to my mind at least) belongs together. If I had not created the multilang package, something like the utils package would be full of multilang code that both the ShellBolt and ShellSpout use. I suppose I could still create the multilang package and place all new code in there and leave the old code where it was. It means things are a bit all over the place, but at least that won't break anything. I do see these fragmented packages becoming unwieldy in the future.

xumingming commented:
Yeah, understand your concern. I personally like the other option your suggested: all new code goes into the new package.

Commit:
Moved ShellBolt, ShellProcess and ShellSpout back to their old packages. Updated dependencies.

jsgilmore commented:
I've now moved the existing code back to where it was and updated the dependancies. With the new code in the new package.

xumingming commented on project.clj:
```diff
@@ -10,7 +10,7 @@
                  :post "storm-user@googlegroups.com"}
   :dependencies [~@DEPENDENCIES]
-  :plugins [[~'lein-sub "0.2.1"]]  
+  :plugins [[~'lein-sub "0.2.1"],[no-man-is-an-island/lein-eclipse "2.0.0"]]
```
better keep the plugin config `lein-eclipse` in your own `profiles.clj` (`lein-sub` is configured here because it is kinda of a must due to current storm's multi-module structure)

jsgilmore replied:
Fixed

xumingming commented on storm-core/src/clj/backtype/storm/bootstrap.clj:
```diff
@@ -17,7 +17,7 @@
      (import (quote [backtype.storm.task IBolt IOutputCollector
                      OutputCollector TopologyContext ShellBolt
                      GeneralTopologyContext WorkerTopologyContext]))
-     (import (quote [backtype.storm.coordination CoordinatedBolt CoordinatedBolt$SourceArgs 
```
remove white space.

jsgilmore replied:
Sorry, I couldn't find white space in the new code.

xumingming commented:
better add some comments for the new codes, e.g. `Emission`, `Immission`, it makes it easier for the future maintainer to understand the code.

jsgilmore commented:
Will do. Referring to what @francis-liberty mentioned, that there might be other message types, already internal to Storm, which implement the fields of Immission and Emission? Could @xumingming  comment on that? Perhaps the tuple message type? I would prefer to use internal message types, if they already exist.

Commit:
Added comments to the new multilang objects.

xumingming commented on storm-core/src/jvm/backtype/storm/multilang/JsonSerializer.java:
```diff
@@ -0,0 +1,147 @@
+	}
+
+	public Number connect (Map conf, TopologyContext context) throws IOException, NoOutputException {
+		JSONObject setupInfo = new JSONObject();
```
seems the code is not well formatted, used `tab` in code?

jsgilmore replied:
I did. Should I use spaces? Are there any guidelines available for how you guys format your code?

xumingming replied:
generally use 4 spaces to indent rather than a tab.

xumingming commented on storm-core/src/jvm/backtype/storm/spout/ShellSpout.java:
```diff
@@ -18,88 +21,82 @@
     private SpoutOutputCollector _collector;
     private String[] _command;
     private ShellProcess _process;
+    private SpoutMsg spoutMsg;
```
we should have three `SpoutMsg` object to represents the three message types: `fail`, `ack`, `next` commands(rather than just use a single one) -- these messages can come concurrently.

jsgilmore replied:
I wasn't aware of that. What would then happen if two ack messages come concurrently? Otherwise, how does Storm know when a spout has finished processing an ack?

xumingming replied:
I mean a `fail` message and a `next` message can come concurrently, so we can not use a single object to represent both of them.

jsgilmore replied:
Looking at the ShellSpout implementation, I can't see how this can work. If a spout receives a next message, it calls querySubprocess(). This function is a loop that sends a single message to the non-JVM program attached to it and then waits to receive messages until the non-JVM program sends a sync.

If next and ack can be received concurrently from Storm, it means that there can be two execution threads in the querySubprocess function at the same time. This would mean that the non-JVM program can receive an ack message, even while it is sending messages as a response to a next message. It is explicitly stated in the Storm multilang protocol that this can't happen.

xumingming replied:
If I understand it correctly, this issue is not multilang protocol related. `ShellSpout` is a normal Storm spout, so the method `fail`, `nextTuple`, `ack` may be invoked by storm-core concurrently, so it seems that we need to synchronize the call to method `querySubprocess`? how do you think @nathanmarz ?

nathanmarz replied:
Spout methods are never called concurrently by Storm. Any given spout object (remember the spout object is duplicated for each task that makes up the spout around the cluster) has all methods on it called from the same thread.

xumingming replied:
thanks Nathan, then it's ok to have one SpoutMsg for all three message types. But personally I still think use a single SpoutMsg and change its property every time seems a little tricky.

xumingming commented:
A new config item should be added to `Config.java` and `defaults.yaml`, so developer can specify a custom serializer.

xumingming commented on storm-core/src/jvm/backtype/storm/utils/ShellProcess.java:
```diff
@@ -1,85 +1,79 @@
 import org.apache.log4j.Logger;
 
-public class ShellProcess {
+public class ShellProcess implements Serializable {
```
Why we need to implement `Serializable`?

jsgilmore replied:
When nimbus has to serialize the ShellProcess class to send it to supervisors, it wants it declared as serializable. I know it wasn't declared as serializable before and the only reason I can think of why it's needed now is because it contains an interface as member.

All I know is that topology builder wouldn't allow me to submit a topology. This was under Java 7, if that makes a difference.

xumingming commented on storm-core/src/jvm/backtype/storm/spout/ShellSpout.java:
```diff
@@ -18,88 +21,82 @@
+
+    public ShellSpout(ISerializer serializer, String... command) {
+        _command = command;
+        _process = new ShellProcess(serializer, _command);
```
`new ShellProcess` might need to move to `open` method, so we can first get user configured serializer from `stormConf`, and pass the serializer to the `ShellProcess` constructor

xumingming commented on storm-core/src/jvm/backtype/storm/spout/ShellSpout.java:
```diff
@@ -18,88 +21,82 @@
+        _process = new ShellProcess(new JsonSerializer(), _command);
+    }
+
+    public ShellSpout(ISerializer serializer, String... command) {
```
This constructor seems not needed, since we will get the serializer class from `stormConf`(rather than let developer specify it in code).

jsgilmore replied:
This sounds like a much better solution. I've not worked with stormConf like that before. Will look into how to do that.

xumingming commented on storm-core/src/jvm/backtype/storm/spout/ShellSpout.java:
```diff
@@ -18,88 +21,82 @@
+
     public ShellSpout(String... command) {
         _command = command;
+        _process = new ShellProcess(new JsonSerializer(), _command);
```
hardcoded `JsonSerializer` should not be here, we might have a `default serializer` specified in `defaults.yaml`

xumingming commented on storm-core/src/jvm/backtype/storm/task/ShellBolt.java:
```diff
@@ -57,54 +56,53 @@
+        _process = new ShellProcess(new JsonSerializer(), _command);
+    }
+
+    public ShellBolt(ISerializer serializer, String... command) {
```
just like `ShellSpout`: 
- this constructor is not needed
- `new ShellProcess` should be moved to `prepare` method
- hardcoded `JsonSerializer` should go to `defaults.yaml`

xumingming commented on storm-core/src/jvm/backtype/storm/spout/ShellSpout.java:
```diff
@@ -18,88 +21,82 @@
+                    Object messageId = emission.getId();
+                    if (task == 0) {
                         List<Integer> outtasks = _collector.emit(stream, tuple, messageId);
-                        Object need_task_ids = action.get("need_task_ids");
```
this `need_task_ids` logic is removed?

jsgilmore replied:
I removed this logic, since nowhere in the multilang protocol specification is this specified as an option (https://github.com/nathanmarz/storm/wiki/Multilang-protocol). I can put it back, which will just require an additional field to the Emission object. We then just have to update the multilang specification as well, which I won't mind doing.

jsgilmore replied:
I've re-added this option.

xumingming commented on storm-core/src/jvm/backtype/storm/task/ShellBolt.java:
```diff
@@ -57,54 +56,53 @@
-                        if (action == null) {
-                            // ignore sync
-                        }
+                        Emission emission = _process.readEmission();
```
is it possible that the `emission` is null, just like the old `action`?

jsgilmore replied:
So I looked at this code, and I could find no reason for action to be null as a consequence of a sync message, unless sync were sent only as "end" messages in the past. _process.readMessage (as it stands currently in the Storm code), reads a string and then parses it. If you look at readString, if only an "end" was received readString will return an empty string, which will translate to a null message. But for a sync message of the form: {"command":"sync"}, I can see no reason why action would be null.

The reason why I think this has not affected the use of shell components, is that if a valid action message were received (in the JSON format specified in the multilang protocol), it would also just be ignored, since the "sync" command is never handled in the if else over the "command" variable.

xumingming commented on storm-core/src/jvm/backtype/storm/task/ShellBolt.java:
```diff
@@ -113,19 +111,20 @@ public void run() {
             public void run() {
                 while (_running) {
                     try {
+                    	// FIXME: This can either be TaskIds or Immissions
```
this `FIXME` not fixed?

jsgilmore replied:
Ah, sorry. The if ... instaceof ... fixed that.

xumingming commented:
a minor comment: are the terms `Emission`, `Immission` used frequently in your code? seems not very easy to understand at first glance(my feeling may be wrong, since English is not my native language).

jsgilmore commented:
An emission is something that is received from somewhere, while an immission is something that is sent to somewhere. Admittedly, immission is not a word that is regularly used, although it is a perfectly good English word. It is also technically the opposite of emission, which is why I chose it. As I mentioned earlier, if there are other messages, already part of Storm, that contain the fields of emission and immission, I would rather use them.

If you can think of better names for them, I'm also open to that :-)

francis-shuoch commented:
@jsgilmore the class I recommended to use is Tuple. Here is the link to the [interface](https://github.com/nathanmarz/storm/blob/master/storm-core/src/jvm/backtype/storm/tuple/Tuple.java) and a storm provided [base class](https://github.com/nathanmarz/storm/blob/master/storm-core/src/jvm/backtype/storm/tuple/TupleImpl.java). I think it already has the fields you created in Emission and Immission. You might take a look. ;)

xumingming commented:
@francis-liberty `Tuple` in storm is designed to represent the messages transfered between Spouts and Bolts. I think it's not appropriate to use it here.

jsgilmore commented:
That is how I understood the purpose of tuple. @francis-liberty you'll note that both Emission and Immission have tuple fields. Although they are lists of objects, that represent the idea of a tuple in Storm. In addition to carrying tuples, emissions also contain all required meta data to emit them, but an emission can be any of: "emit", "sync", "log", "ack", "fail" or "error", as specified in the command field.

These emission types specify which actions the shell component will take in response to receiving an emission.

jsgilmore commented:
I could perhaps call an Immission a BoltMsg. Then emissions are messages sent from both spout and bolt non-JVM components to shell components, A SpoutMsg is a message sent from a ShellSpout to a non-JVM spout and a BoltMsg is a message sent from a ShellBolt to a non-JVM bolt.

I could take this pattern further and then call an emission a ShellMsg or a StormMsg, since it is a message sent from a non-JVM component to Storm or a Shell component. So this naming scheme would then be the destination of the message is its name.

xumingming commented:
haha, naming things is the most difficult part of programming. I think the new names you proposed are easier to understand than `Emission` and `Immission`.

Commit:
multilang serializer is not handled by stormConf. Renamed Emission and Immission to ShellMsg and BoltMsg respectively. Documented ISerializer interface.

Commit:
multilang serializer is now handled by stormConf. Renamed Emission and Immission to ShellMsg and BoltMsg respectively. Documented ISerializer interface.

francis-shuoch commented:
I think @jsgilmore 's naming convention is: w.r.t non-JVM component, "e-" means out, "i-" means in. So "emission" is to emit out of non-JVM component, "immission" is to emit in (or inject, actually) to non-JVM component. Is this more clear now?

Commit:
Merge branch 'master' of github.com:jsgilmore/storm

jsgilmore commented:
So I've changed these names now to make them more understandable. But just for interest's sake: you don't have to play with "e-"s and "i-"s. They're just words for what happens from the non-JVM component's perspective.
From the free online dictionary:
e·mis·sion
n.
1. The act or an instance of emitting.
2. Something emitted.
http://www.thefreedictionary.com/Emission

Im`mis´sion
n.  1.  The act of immitting, or of sending or thrusting in; injection; - the correlative of emission.
http://www.thefreedictionary.com/immission

francis-shuoch commented:
Ok, thanks for these tips ;). Good job ~~

xumingming commented:
also, add some unit tests for this new feature.

jsgilmore commented:
My editor removes trailing whitespace. I thought this was required. The editor removed trailing whitespace in Config.java (of which there was a lot).

I couldn't really find shell component tests. backtype.storm.testing seems to have unit tests that make use of Storm's unit testing API for spouts and bolt. I couldn't really find any Storm component unit tests. How do you guys do unit tests in Storm? Where should my tests live? What do you use the run the tests (Maven?)?

Thanks for all the great feedback!

xumingming commented:
tests are placed in `storm-core/test/clj/backtype/storm`, tests are all writen in clojure, and run the tests with `lein test`

Commit:
Merge branch 'master' of https://github.com/nathanmarz/storm

Commit:
Added newline to NoOutputException

jsgilmore commented:
I've had a look at the Storm tests. Specifically the shell component tests for ruby and python (multilang_test.clj). These tests also test the changes I made to Storm. They start up shell components and those shell components start to communicate. I didn't really add a feature to Storm, merely refactored code to be more modular.

Commit:
Re-added need_task_ids option

Commit:
Added json msg and id parsing.

Commit:
Allow for null boltMsg when polling

xumingming commented:
Cool! I think we are almost there except that those pure whitespace-changes need to be removed(`project.clj`, `bootstrap.clj`, `Config.java`).

Commit:
Removed pure whitespace changes of project.clj and bootstrap.clj

jsgilmore commented:
So I've reset the pure whitespace changes to project.clj and bootstrap.clj. Config.java was not a pure whitespace change; I added the STORM_MULTILANG_SERIALIZER item.

Just say if want me to re-add all the trailing whitespace in Config.java if that's what's required for the PR to be accepted :-)

xumingming commented:
For Config.java, I think the easiest way is to revert the current changes and add your new STORM_MULTIANG_SERIALIZER back.

Commit:
Added some trailing whitespace to Config.java

jsgilmore commented:
Done

xumingming commented:
+1 for adding this PR into `0.9.1`

jsgilmore commented:
Sorry, I don't follow. Are you asking me to merge something more from upstream? I just now merged upstream master and there doesn't seem to be any changes and that's still on 0.9.0.

xumingming commented:
Oh, I think there is nothing more you need to do :) I am just voting to add your PR to the next release. Current release(`0.9.0`) only accept bugfix PR now.

Commit:
Merge branch 'master' of https://github.com/nathanmarz/storm

Commit:
Merge branch 'master' of https://github.com/nathanmarz/storm

Commit:
Added debugging info to ShellSpout and pulled in latest upstream master.

Commit:
Added null check for anchors. Log a component's error stream under it name.

ptgoetz committed:
align with version 0.9.0.1, clean up m2-pom.xml and align with project.clj, update README

miguno committed:
Merge pull request #50 from ptgoetz/master

Align storm-starter with upstream Storm 0.9.0 (release).

Commit:
m2-pom.xml: fix indentation / whitespace

Commit:
README: Add instructions for IntelliJ IDEA

wurstmeister committed:
updated to final version of kafka 0.8

Commit:
Re-add Google Guava to maven deps config (accidentally removed in commit 8012cd7)

Commit:
AbstractRankerBolt: Emit defensive copy of Rankings object

This addresses ConcurrentModificationException that might be thrown in
the ranker bolts, when the Rankings object would be modified (by the
emitting bolt) after sending it downstream.

Commit:
README: Fix instructions for IntelliJ IDEA

wurstmeister committed:
updated to final version of storm 0.9.0

wurstmeister committed:
preparing for tag

wurstmeister committed:
added distribution management

* bumped version
* updated group id

wurstmeister committed:
updated group id, prepared for tag

wurstmeister committed:
updated readme

wurstmeister committed:
use consistent formatting

wurstmeister committed:
removed empty statement

wurstmeister committed:
Renamed HostPort to Broker

wurstmeister committed:
Use partition id in zk path

* added topology id to logs

wurstmeister committed:
added changelog

wurstmeister committed:
Update README.md

wurstmeister committed:
Added error handling for fetch request

* allow retry of fetch if offset was invalid
* updated changelog

Commit:
Do not rely on spaces to separate command args

Commit:
URL-decode special chars in topo ids

wurstmeister committed:
added OFFSET_METADATA_TOO_LARGE error

Commit:
supervisor logs worker command with '-quoted tokens

Commit:
Do not encode,decode topo id when launching worker

Commit:
Support *.worker.childopts as list or as string

Commit:
Add test for handling *.worker.childopts in supvor

wurstmeister committed:
calculate start offset for new topology consistently

wurstmeister committed:
calculate start offset for new topology consistently

EdisonXu committed:
fix trident keep refreshing partition issue

wurstmeister committed:
Merge pull request #13 from EdisonXu/fix1

fix keep refreshing partitions issue

wurstmeister committed:
use guava equals/hashCode

wurstmeister committed:
added profile for scala 2.10 build

yerenkow committed:
Fix for launching jar when some JVM opts are specified, now corrected and using shlex library

Commit:
STORM-70 Upgrade to ZK-3.4.5 and curator-1.3.3

wurstmeister committed:
preparing for tag

wurstmeister committed:
update to next snapshot version

ptgoetz committed:
[maven-release-plugin] prepare release v0.9.1-incubating

ptgoetz committed:
[maven-release-plugin] prepare for next development iteration

Commit:
Make the spout's Storm Metrics' time bucket size configurable via
TridentKafkaConfig. Leave the default time bucket size at 60 seconds.

ptgoetz committed:
remove Nathan Marz reference from NOTICE file

ptgoetz committed:
undo last commit

anfeng committed:
remove Yahoo copyright from NOTICE files

supercargo committed:
STORM-241: making thrift binary chunk size configurable

dschiavu committed:
Update pom.xml

Revert custom changes

anfeng committed:
NOTICE revised regarding Yahoo contribution

xiaokang committed:
STORM-186 fix float secs to millis long convertion

wurstmeister committed:
Merge pull request #22 from dschiavu/configurable-metrics-emit-interval

Make the spout's Storm metrics time bucket size configurable via TridentKafkaConfig. Leave the default time bucket size at 60 seconds.

wurstmeister committed:
use metricsTimeBucketSizeInSecs consistently

xiaokang committed:
STORM-186 convert 1000 to long to avoid int overflow

Commit:
Add a new per-partition and total metric, "earliestTimeOffset". Rename
"latestTime" metric to "latestTimeOffset" to more closely conform to
other, related metrics' nomenclature.

wurstmeister committed:
Added support for reading message key

wurstmeister committed:
fixed tests

wurstmeister committed:
preparing for tag

wurstmeister committed:
added Kafka bolt

wurstmeister committed:
Merge pull request #25 from dschiavu/new-metric-earliest-offset

Add a new per-partition and total metric, "earliestTimeOffset"

Commit:
Work around ZOOKEEPER-1476 by prefering IPv4 by default.

supercargo committed:
Moving chunk size out of configuration into private constant

supercargo committed:
whitespace cleanup

Commit:
Fixing an integer overflow in getSleepTimeMs().  This should fix the
issue of negative values passed to sleep, causing
IllegalArgumentExceptions.

ptgoetz committed:
update CHANGELOG

ptgoetz committed:
Merge branch 'STORM-186' of https://github.com/xiaokang/incubator-storm

ptgoetz committed:
update CHANGELOG

ptgoetz committed:
merge storm-starter into examples

ptgoetz committed:
add examples/storm-starter to build

ptgoetz committed:
add Apache license headers

ptgoetz committed:
update README to reflect incorporation into storm source tree

nathanmarz committed:
remove nathan from notices

anfeng committed:
Merge pull request #1 from apache/master

merge from apache/incubator-storm

ptgoetz committed:
update commons-io dependency version

ptgoetz committed:
Merge branch 'storm-70-zk-upgrade' of github.com:revans2/incubator-storm

ptgoetz committed:
Merge branch 'master' of https://git-wip-us.apache.org/repos/asf/incubator-storm

ptgoetz committed:
Merge branch 'master' of github.com:anfeng/incubator-storm

ptgoetz committed:
fix spacing issue

ptgoetz committed:
add examples source code to binary distribution

ptgoetz committed:
update changelog

ptgoetz committed:
update changelog

nathanmarz committed:
adds capability to remove keys from memorymapstate (maintains proper opaque semantics)

ptgoetz committed:
Merge remote-tracking branch 'original/master' into merge

ptgoetz committed:
update storm to 0.9.1-incubating

ptgoetz committed:
Merge branch 'master' of github.com:wurstmeister/storm-kafka-0.8-plus

supercargo committed:
Merge branch 'master' of https://github.com/apache/incubator-storm into storm-241

ptgoetz committed:
align with 0.9.2-incubating dependencies.

ptgoetz committed:
Merge branch 'storm-starter'

ptgoetz committed:
update README with new committers

ptgoetz committed:
STORM-259

apetresc committed:
Fix race condition in Time.java

Some of my test runs were occasionally failing with a NullPointerException on `backtype.storm.utils.Time.java:64`.

After a bit of investigation, it seems there's a race condition here; if we disable simulating mode while a thread is currently sleeping, then when it wakes up it won't re-check if it's still in "simulating" mode, it'll try to remove the sleep time, and get the NPE.

Refs [STORM-260](https://issues.apache.org/jira/browse/STORM-260).

nielsbasjes committed:
Bolt with whitespace in name can be viewed in ui

miguno committed:
storm-starter: fix base directory path to point to `incubator-storm`

miguno committed:
storm-starter: remove leiningen section because leiningen is not used anymore

nielsbasjes committed:
Ensure that no evil chars can be put into the HTML via the the name of a component

nielsbasjes committed:
Ensure that also the topology id is encoded/decoded correctly

nathanmarz committed:
Merge branch 'removable-memory-map-state'

ptgoetz committed:
STORM-263: update to carbonite 1.4.0/kryo 2.21

wurstmeister committed:
fixed KafkaSpout partition assignment

* added partitions -> task mapping logging

wurstmeister committed:
upgraded storm to 0.9.1

Commit:
STORM-12 Reduce thread usage of Netty transport.

Commit:
STORM-270 don't package .clj files in release jars.

wurstmeister committed:
Update README.md

added yourkit acknowledgement

Commit:
STORM-273. Error while running storm topologies on Windows using "storm jar"

Commit:
Add STORM-263 to changelog.

wurstmeister committed:
close resources in tests

wurstmeister committed:
updated log messages

wurstmeister committed:
removed duplicated code

revans2 committed:
STROM-247: Replace links to github resources in storm script (for Julien Nioche)

revans2 committed:
Merge branch 'master' of https://github.com/sureshms/incubator-storm

STORM-273: Error while running storm topologies on Windows using "storm jar"

revans2 committed:
STORM-273 updated change log.

ptgoetz committed:
Merge branch 'master' of github.com:wurstmeister/storm-kafka-0.8-plus

ptgoetz committed:
Merge branch 'master' of /Users/tgoetz/Projects/apache/../other/storm-kafka-0.8-plus

ptgoetz committed:
Add storm-kafka as an external module.

revans2 committed:
Merge branch 'no-clj-in-jar' of https://github.com/revans2/incubator-storm

STORM-270 don't package .clj files in release jars.

revans2 committed:
Added STORM-270 to changelog

ptgoetz committed:
update to kafka 0.8.1

ptgoetz committed:
add topic to committed path

ptgoetz committed:
update README

ptgoetz committed:
apply Apache license headers

ptgoetz committed:
Merge branch 'STORM-258'

ptgoetz committed:
Merge branch 'master' of https://git-wip-us.apache.org/repos/asf/incubator-storm

ptgoetz committed:
STORM-258: update changelog

ptgoetz committed:
Merge branch 'patch-1' of github.com:apetresc/incubator-storm

ptgoetz committed:
update CHANGELOG

ptgoetz committed:
add JIRA # to changelog entry

ptgoetz committed:
Merge branch 'readme-base-dir' of github.com:miguno/incubator-storm

ptgoetz committed:
Merge branch 'readme-leiningen' of github.com:miguno/incubator-storm

ptgoetz committed:
Merge branch 'master' of github.com:nielsbasjes/incubator-storm

Commit:
updated my e-mail address

Commit:
clean up fd after reading config

yerenkow committed:
Remove too verbose comment about JAR_JVM_OPTS being array

revans2 committed:
Merge branch 'fix-storm-shlex' of https://github.com/yerenkow/incubator-storm
STORM-196: When JVM_OPTS are set, storm jar fails to detect storm.jar from environment

revans2 committed:
Updated Changelog STORM-196: When JVM_OPTS are set, storm jar fails to detect storm.jar from environment

ptgoetz committed:
Merge branch 'fix-readconfig-fd-leak' of github.com:d2r/incubator-storm

ptgoetz committed:
STORM-281: update changelog

ptgoetz committed:
Merge branch 'netty-thread-usage' of github.com:revans2/incubator-storm

ptgoetz committed:
STORM-12: update changelog

Commit:
Treat command line "-c" option number config values as such (STORM-173)

Commit:
Merge remote-tracking branch 'apache/master' into d2r-worker-lancher-args

Conflicts:
	pom.xml
	storm-core/src/clj/backtype/storm/ui/core.clj

Commit:
escape single quotes when logging

bbaugher committed:
Fixed storm-core shade plugin config

Commit:
Move command line options values introspection logic to
Utils#readCommandLineOpts (STORM-173)

Commit:
Parse command line options as JSON values (STORM-173)

Commit:
Clean up imports (STORM-173)

Commit:
STORM-288 : Fixes version spelling

revans2 committed:
Added STORM-285 to Changelog.

ptgoetz committed:
add storm-kafka files to the binary distribution

ptgoetz committed:
Merge branch 'storm-kafka-module'

ptgoetz committed:
STORM-208: update changelog

revans2 committed:
Merge branch 'master' of https://github.com/srinivasgumdelli/incubator-storm

revans2 committed:
Added STORM-288: Fixes version spelling in pom.xml

revans2 committed:
Merge branch 'd2r-worker-lancher-args' of https://github.com/d2r/incubator-storm
STORM-194: Support list of strings in *.worker.childopts, handle spaces

revans2 committed:
Updated Changelog for STORM-194

revans2 committed:
Merge branch 'STORM-173-storm-number-config-value' of https://github.com/dschiavu/incubator-storm
STORM-173: Treat command line "-c" option number config values as such

revans2 committed:
Updated Changelog for STORM-173

Commit:
Merge branch 'master' of https://github.com/nathanmarz/storm

Conflicts:
	conf/defaults.yaml

Commit:
Merge branch 'master' of github.com:jsgilmore/incubator-storm
